File: classical_mechanics/double_pendulum/cpp/tests/test_numerics.cpp
#include "numerics.hpp"

#include <cassert>
#include <cmath>
#include <cstdio>
#include <vector>

// Minimal test harness (no external deps)
static int failures = 0;

#define CHECK_CLOSE(actual, expected, tol)                                          \
    do {                                                                            \
        double _a = (actual), _e = (expected), _t = (tol);                         \
        if (std::abs(_a - _e) > _t) {                                               \
            std::printf("FAIL [%s:%d]  got %.10g  expected %.10g  (tol %.2e)\n",   \
                        __FILE__, __LINE__, _a, _e, _t);                            \
            ++failures;                                                             \
        } else {                                                                    \
            std::printf("PASS  got %.10g\n", _a);                                  \
        }                                                                           \
    } while (0)

int main() {
    // --- integrate_trapezoid ---
    {
        // integral of 1 over [0,1] == 1
        const int N = 1000;
        std::vector<double> x(N), f(N, 1.0);
        for (int i = 0; i < N; ++i) x[i] = static_cast<double>(i) / (N - 1);
        CHECK_CLOSE(mathphys::integrate_trapezoid(f, x), 1.0, 1e-9);
    }
    {
        // integral of x^2 over [0,1] == 1/3
        const int N = 10000;
        std::vector<double> x(N), f(N);
        for (int i = 0; i < N; ++i) {
            x[i] = static_cast<double>(i) / (N - 1);
            f[i] = x[i] * x[i];
        }
        CHECK_CLOSE(mathphys::integrate_trapezoid(f, x), 1.0 / 3.0, 1e-5);
    }

    // --- finite_difference (order 1) ---
    {
        // derivative of 3.7*x + 1.5 == 3.7
        const int N = 100;
        std::vector<double> x(N), f(N);
        for (int i = 0; i < N; ++i) {
            x[i] = static_cast<double>(i) / (N - 1);
            f[i] = 3.7 * x[i] + 1.5;
        }
        auto df = mathphys::finite_difference(f, x, 1);
        CHECK_CLOSE(df[N / 2], 3.7, 1e-9);
    }

    if (failures == 0) {
        std::printf("\nAll tests passed.\n");
        return 0;
    }
    std::printf("\n%d test(s) FAILED.\n", failures);
    return 1;
}

File: classical_mechanics/double_pendulum/python/tests/test_numerics.py
"""Tests for mathphys.numerics."""

import numpy as np
import pytest

from mathphys.numerics import integrate_trapezoid, finite_difference


def test_integrate_trapezoid_constant():
    """Integral of 1 over [0, 1] should be 1."""
    x = np.linspace(0, 1, 1000)
    f = np.ones_like(x)
    result = integrate_trapezoid(f, x)
    assert abs(result - 1.0) < 1e-6


def test_integrate_trapezoid_sin():
    """Integral of sin(x) over [0, pi] should be 2."""
    x = np.linspace(0, np.pi, 10_000)
    f = np.sin(x)
    result = integrate_trapezoid(f, x)
    assert abs(result - 2.0) < 1e-5


def test_integrate_trapezoid_x_squared():
    """Integral of x^2 over [0, 1] should be 1/3."""
    x = np.linspace(0, 1, 10_000)
    result = integrate_trapezoid(x**2, x)
    assert abs(result - 1.0 / 3.0) < 1e-5


def test_finite_difference_linear():
    """Derivative of a linear function should be its slope."""
    x = np.linspace(0, 1, 100)
    slope = 3.7
    f = slope * x + 1.5
    df = finite_difference(f, x, order=1)
    # Interior points should match slope exactly for linear functions
    assert np.allclose(df[1:-1], slope, atol=1e-10)


def test_finite_difference_quadratic_second_order():
    """Second derivative of x^2 should be 2."""
    x = np.linspace(0, 1, 1000)
    f = x**2
    d2f = finite_difference(f, x, order=2)
    assert np.allclose(d2f[2:-2], 2.0, atol=1e-4)


def test_finite_difference_invalid_order():
    x = np.linspace(0, 1, 10)
    with pytest.raises(ValueError, match="Unsupported derivative order"):
        finite_difference(x, x, order=3)
